elastic/kibana
# [Fleet] Add experimental package verification (server side)

## Summary

Part of #133822.

Add package verification on package install. Verify a package signature if:
- the feature flag is enabled `packageVerification`
- the package has a .sig file in the registry (none will yet)
- a local public key is configured at `xpack.fleet.packageVerification.gpgKeyPath`

Otherwise package verification status is always set to "unknown".

### Summary of changes

- Add new feature flag `packageVerification`, disabled by default
- Add new fields to installation saved object (saved object migration to follow in separate PR)
    - `verification_status` - `unknown | verified | unverified`
    - `verification_key_id` - present if status is not unknown, they id of the key used to do verification.
- Add new cache for storing archive verification status
- When downloading archive, perform verification and cache result
- Verify bundled packages when downloading them (but don't currently store the .sig file for future verification, so will still have unknown verification status)
- Return 400 response code if package verification fails
- If a package is installed with `force: true` then install will proceed even if verification fails

### Testing steps

- 1. Run the version of the EPR with package signatures locally:

```docker run -p 8080:8080 docker.elastic.co/observability-ci/package-registry/distribution:PR-463```
- 2. set `xpack.fleet.registryUrl: http://localhost:8080`
- 3. download local public key `curl https://artifacts.elastic.co/GPG-KEY-elasticsearch -o /tmp/GPG-KEY-elasticsearch  `
- 4. set `xpack.fleet.packageVerification.gpgKeyPath: /tmp/GPG-KEY-elasticsearch`
- 5. Enable the feature by setting:

```
xpack.fleet.enableExperimental: 
    - packageVerification
```

- 6. Optional - to test if verification fails, here is a GPG public key which isnt the elastic one: 
[NOT_THE_ELASTIC_KEY.txt](https://github.com/elastic/kibana/files/9021435/NOT_THE_ELASTIC_KEY.txt)

Test scenarios:

- **if verification feature flag is enabled and**:
    - **verification is successful**: Install should succeed,`verification_key_id` should be `d27d666cd88e42b4` and `verification_status` should be `verified`
    - **verification is not successful** : (e.g by using a different public key supplied above)  Install should not succeed, 400 response code returned,`verification_key_id` should be `d27d666cd88e42b4` and `verification_status` should be `unverified`
    - **no public key configured locally**: (e.g by unsetting gpgKey config)Install should succeed,`verification_key_id` should not be set and `verification_status` should be `unknown`
    - **missing .sig file**: (e.g by connecting to the normal EPR) Install should succeed,`verification_key_id` should not be set and `verification_status` should be `unknown`
- **verification feature flag not enabled**: Install should succeed,`verification_key_id` should not be set and `verification_status` should be `unknown`


### Checklist

Delete any items that are not applicable to this PR.
